ios/versioned/sdk48/EXFileSystem/EXFileSystem/ABI48_0_0EXFileSystem.m

```diff
@@ -269,13 +269,13 @@ - (NSDictionary *)encodingMap
 }
 
 ABI48_0_0EX_EXPORT_METHOD_AS(writeAsStringAsync,
-                    writeAsStringAsyncWithURI:(NSString *)uriString
+                    writeAsStringAsyncWithURI:(NSString *)pathString
                     withString:(NSString *)string
                     withOptions:(NSDictionary *)options
                     resolver:(ABI48_0_0EXPromiseResolveBlock)resolve
                     rejecter:(ABI48_0_0EXPromiseRejectBlock)reject)
 {
-  NSURL *uri = [NSURL URLWithString:uriString];
+  NSURL *uri = [NSURL URLWithString:[pathString stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]]];
```

Comment:
Should this use `URLPathAllowedCharacterSet` since the intent is to encode a path rather than a query string?

Reply:
I agree, and would also think we should be using `[NSURL fileURLWithString]` maybe

Reply:
Unfortunately using `[NSURL fileURLWithPath]` directly does not work, it produces the same `Error: File 'file:/Users/.../%E4%B8%AD%E6%96%87 -- file:///' isn't readable.` error.

Reply:
I did some more investigation on this and it seems that the problem is related to the way that the NSURL path is returned.

Given the following path, `FileSystem.documentDirectory + 'ÅÄÖ'` (`file:///var/mobile/Containers/Data/Application/6FAE142B-76F6-4FDD-AE26-A70292E956B9/Documents/ÅÄÖ`) which does not exist in the device:

- Using `NSURL URLWithString:[pathString stringByAddingPercentEncodingWithAllowedCharacters`:
    - `NSCharacterSet URLQueryAllowedCharacterSet` produces `/var/mobile/Containers/Data/Application/6FAE142B-76F6-4FDD-AE26-A70292E956B9/Documents/ÅÄÖ` (does not include `file` in the path) and scheme is `file`
   - Using `NSCharacterSet URLPathAllowedCharacterSet` produces `file:///var/mobile/Containers/Data/Application/6FAE142B-76F6-4FDD-AE26-A70292E956B9/Documents/ÅÄÖ` (same as input) and scheme is `(null)`
- Using `[NSURL fileURLWithPath:pathString]` produces `/file:/var/mobile/Containers/Data/Application/6FAE142B-76F6-4FDD-AE26-A70292E956B9/Documents/ÅÄÖ` (removes ///) and scheme is `file`

Whenever the returned scheme is `(null)` we already fallback to `[NSURL fileURLWithPath:pathString isDirectory:false]` producing `/file:/var/mobile/Containers/Data/Application/D886A047-E4B6-4365-A5FF-04372FF22182/Documents/ÅÄÖ`
https://github.com/expo/expo/blob/fa1f8b5219dea4ae2c243c4bbff4577443a0952b/packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m#L184-L186

After that, we check for permissions on the given URI
https://github.com/expo/expo/blob/fa1f8b5219dea4ae2c243c4bbff4577443a0952b/packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m#L1003-L1005

And here is the problem, inside the `getInternalPathPermissions` function https://github.com/expo/expo/blob/fa1f8b5219dea4ae2c243c4bbff4577443a0952b/packages/expo-file-system/ios/EXFileSystem/EXFilePermissionModule.m#L30-L48


we check specifically for `standardizedPath hasPrefix:[scopedDirectory stringByAppendingString:@"/"]` which is not true because our `standardizedPath` starts with `/file:`. To solve that we could add a check for `[standardizedPath hasPrefix:[NSString stringWithFormat:@"/file:%@/", scopedDirectory]] ||`. What do you guys think about this?